pkg/util/exec/exec.go

```diff
@@ -322,10 +329,27 @@ func assertErrorType(err error) string {
 }
 
 // ExtractExitCode attempts to get the exit code from the error returned by an Executor function.
-// This should also work for any errors returned by the golang os/exec package.
+// This should also work for any errors returned by the golang os/exec package and "k8s.io/utils/exec"
 func ExtractExitCode(err error) (int, error) {
-	if ee, ok := err.(*exec.ExitError); ok {
-		return ee.ExitCode(), nil
+	switch errType := err.(type) {
+	case *exec.ExitError:
+		return errType.ExitCode(), nil
+
+	case *kexec.CodeExitError:
+		return errType.ExitStatus(), nil
+
+	case *kerrors.StatusError:
+		return int(errType.ErrStatus.Code), nil
+
+	default:
+		logger.Debugf(err.Error())
+		// This is ugly but I don't know why the type assertion does not work...
+		// Whatever I've tried I can see the type "exec.CodeExitError" but none of the "case" nor other attempts with "errors.As()" worked :(
+		// So I'm parsing the Error string until we have a solution
+		if strings.Contains(err.Error(), "command terminated with exit code") {
+			a := strings.SplitAfter(err.Error(), "command terminated with exit code")
+			return strconv.Atoi(strings.TrimSpace(a[1]))
+		}
+		return 0, errors.Errorf("error %#v is not an ExitError nor CodeExitError but is %v", err, reflect.TypeOf(err))
 	}
```

Comment:
Maybe it's fundamentally not possible to do this with a switch on the type given that the type may be partly obfuscated in Go.

As an example, this is from the k8s codebase:
```go
// IsNotFound returns true if the specified error was created by NewNotFound.
// It supports wrapped errors.
func IsNotFound(err error) bool {
	return ReasonForError(err) == metav1.StatusReasonNotFound
}

// ReasonForError returns the HTTP status for a particular error.
// It supports wrapped errors.
func ReasonForError(err error) metav1.StatusReason {
	if status := APIStatus(nil); errors.As(err, &status) {
		return status.Status().Reason
	}
	return metav1.StatusReasonUnknown
}
```

See also: https://github.com/golang/go/issues/35874#issuecomment-560903907


And this has some possible hints: https://golang.org/src/os/exec/exec_test.go?h=Error
```go
func TestCatGoodAndBadFile(t *testing.T) {

	// Testing combined output and error values.

	bs, err := helperCommand(t, "cat", "/bogus/file.foo", "exec_test.go").CombinedOutput()

	if _, ok := err.(*exec.ExitError); !ok {

		t.Errorf("expected *exec.ExitError from cat combined; got %T: %v", err, err)

	}
```

So I wonder if something like this would work...?
```go
switch err {
case e, ok := err.(*exec.ExitError); ok:
    return e.ExitCode(), nil
case e, ok := err.(*kexec.CodeExitError); ok:
    return e.ExitStatus(), nil
/// etc.
}

Reply:
Actually, type switches should work based on a test I just did. I

https://play.golang.org/p/ueAvRXpMU0x
This times out in the playground, but it works in a local test.

```
---
*errors.StatusError
kerrors.StatusError

---
*exec.ExitError
exec.ExitError

---
*errors.StatusError
kerrors.StatusError

---
*errors.StatusError
kerrors.StatusError

---
*exec.CodeExitError
kexec.CodeExitError

---
*exec.ExitError
exec.ExitError

---
*exec.CodeExitError
kexec.CodeExitError

---
*errors.StatusError
kerrors.StatusError
```

Reply:
We do check type with `switch` at several places though and it is working.

Reply:
Unfortunately, this:

```go
	if e, ok := err.(*exec.ExitError); ok {
		return e.ExitCode(), nil
	} else if e, ok := err.(*kexec.CodeExitError); ok {
		return e.ExitStatus(), nil
	} else if e, ok := err.(*kerrors.StatusError); ok {
		return int(e.ErrStatus.Code), nil

	} else {
		return 0, errors.Errorf("error %#v is not an ExitError nor CodeExitError but is %v", err, reflect.TypeOf(err))
	}
```


Resulted in the same error:

```
2021-07-07 16:58:01.806942 E | ceph-object-controller: failed to determine return code of 'radosgw-admin' command. assuming this could be an invalid flag error. error exec.CodeExitError{Err:(*errors.errorString)(0xc001123810), Code:22} is not an ExitError nor CodeExitError but is exec.CodeExitError
```

Reply:
Based on the output from `reflect.TypeOf(err)`, it seems like the error isn't a pointer. I think maybe the right thing to do is to check if it's a pointer first, then cast it to a pointer if it isn't? https://stackoverflow.com/questions/36890896/in-go-how-do-i-check-a-value-is-of-any-pointer-type

Reply:
Though it works for me in my test if I include a non-pointer switch case like here: https://play.golang.org/p/x8Zc2q8iSWv

```
---
Type: exec.CodeExitError
TypeOf: exec.CodeExitError
case: kexec.CodeExitError

---
Type: *errors.StatusError
TypeOf: *errors.StatusError
case: *kerrors.StatusError

---
Type: *errors.StatusError
TypeOf: *errors.StatusError
case: *kerrors.StatusError

---
Type: *errors.StatusError
TypeOf: *errors.StatusError
case: *kerrors.StatusError

---
Type: exec.CodeExitError
TypeOf: exec.CodeExitError
case: kexec.CodeExitError

---
Type: *exec.CodeExitError
TypeOf: *exec.CodeExitError
case: *kexec.CodeExitError

---
Type: exec.CodeExitError
TypeOf: exec.CodeExitError
case: kexec.CodeExitError

---
Type: *exec.ExitError
TypeOf: *exec.ExitError
case: *exec.ExitError
```

Reply:
Honestly, I don't know, I hope it's something dumb but I swear I've tried already not using pointer but just for the sake of it... I tried again with this code:

```go
	switch errType := err.(type) {
	case *exec.ExitError:
		return errType.ExitCode(), nil

	case *kexec.CodeExitError:
		return errType.ExitStatus(), nil

	case kexec.CodeExitError:
		return errType.ExitStatus(), nil

	case *kerrors.StatusError:
		return int(errType.ErrStatus.Code), nil

	default:
		logger.Debugf(err.Error())
		// // This is ugly but I don't know why the type assertion does not work...
		// // Whatever I've tried I can see the type "exec.CodeExitError" but none of the "case" nor other attempts with "errors.As()" worked :(
		// // So I'm parsing the Error string until we have a solution
		// if strings.Contains(err.Error(), "command terminated with exit code") {
		// 	a := strings.SplitAfter(err.Error(), "command terminated with exit code")
		// 	return strconv.Atoi(strings.TrimSpace(a[1]))
		// }
		return 0, errors.Errorf("error %#v is not the type we are looking for %v", err, reflect.TypeOf(err))
	}
```


but still got the same issue:

```
2021-07-07 19:24:44.191189 E | ceph-object-controller: failed to determine return code of 'radosgw-admin' command. assuming this could be an invalid flag error. error exec.CodeExitError{Err:(*errors.errorString)(0xc0014f3d40), Code:22} is not an ExitError nor CodeExitError but is exec.CodeExitError
```

Reply:
Opened https://github.com/rook/rook/issues/8280 to track this.